api/test_models.py

```diff
@@ -96,6 +97,29 @@ def test_field_report_create(self):
         self.assertEqual(obj.event, event)
         self.assertIsNotNone(obj.report_date)
 
+    def mocked_requests_post(*args, **kwargs):
+        class MockResponse:
+            def __init__(self, json_data, status_code):
+                self.json_data = json_data
+                self.text       = 'FindThisGUID'
+                self.status_code = status_code
+
+            def json(self):
+                return self.json_data
+
+        return MockResponse(None, 200)
+
+    @mock.patch('utils.erp.requests.post', side_effect=mocked_requests_post)
+    def test_ERP_related_field_report(self, mocked_requests_post):
+        dtype = models.DisasterType.objects.get(pk=1)
+        country = models.Country.objects.create(name='country', iso='CH')
+        event = models.Event.objects.create(name='disaster2', summary='test disaster 2', dtype=dtype)
+        event.countries.set([country])
+        report = models.FieldReport.objects.create(rid='test2', event=event, dtype=dtype, ns_request_assistance=True)
+        self.assertEqual(mocked_requests_post.called, True)
+        ERP = models.ERPGUID.objects.get(api_guid='FindThisGUID')
+        self.assertEqual(ERP.field_report_id, report.id)
```

Comment:
For `Model.objects.create(` let's use factory models instead.

Reply:
I created a FieldReport factory (it did not exist before, in api), but via factory the field report creation (ns_required_assistance=True) did not call the needed method. So I created the other entities via factory, but the field report is created in the old fashioned way.
Nevertheless it was a good practice to crate a new (complicated) factory.

Reply:
This statement will only create a object in memory `fieldReportFactory.FieldReportFactory(rid='test2', dtype=dtype, ns_request_assistance=True)`. You need to use `FieldReportFactory.create(..)` or save report instance to trigger `push_fr_data`.

Reply:
Well, I could not achieve to trigger push_fr_data via the factory way.
It always gave:
api.models.ERPGUID.DoesNotExist: ERPGUID matching query does not exist.
(and if I debug it, models.ERPGUID.objects.all() is really empty.)

If you can fix it, I would greatly appreciate the help.

Reply:
(Hmm... the local test differs from the CircleCI one. Locally self.assertEqual(erp_request_side_effect_mock.called, False) worked, but remotely it failed.)

Reply:
You forgot to pass the event.
I encounter another issue with the factory, when `post_generation` is used, the factory will trigger the model save() **twice** as described [here](https://github.com/FactoryBoy/factory_boy/issues/316). This results in creating **2 ERPGUID**.

If we fix this `TODO: maybe add a check, and only send request if anything has changed`, then it should work as expected.

NOTE: My changes are [here](https://github.com/IFRCGo/go-api/pull/1080/commits/e5586ea765fe95aed2266c74f1853e2cfe1bbffd).

docker-compose exec serve bash -c 'pytest --reuse-db api/test_models.py::FieldReportTest::test_ERP_related_field_report'
![2021-05-25-134909](https://user-images.githubusercontent.com/7059255/119462044-0f87cc00-bd60-11eb-9c61-6ab7a38f30fd.png)

Reply:
I could solve this via commenting out the @factory.post_generation parts in field_report factory.
(The check for double call and allowing only the first seemed to be not a really clear solution.)

In test_utils also used factories.